app/controllers/payments.js

```diff
@@ -1,19 +1,63 @@
 const config = require('config')
 const stripe = require('stripe')(config.stripe.api_secret)
+const plan = config.stripe.enthusiast_plan_id
+const User = require('../models/user.js')
+const logger = require('../../lib/logger.js')()
 
-exports.post = (req, res) => {
-  const amount = 1000
+exports.post = async (req, res) => {
+  let userId
+  let sub
+  let customer
+  try {
+    userId = req.body.userId
+    logger.info(`submitting payment for ${userId}`)
+    const { token: { email, id } } = req.body
 
-  stripe.customers.create({
-    email: req.body.stripeEmail,
-    source: req.body.stripeToken
-  })
-    .then(customer => console.log(customer) ||
-      stripe.charges.create({
-        amount,
-        description: 'Sample Charge',
-        currency: 'usd',
-        customer: customer.id
-      }))
-    .then(charge => console.log(charge) || res.render('main'))
+    customer = await stripe.customers.create({
+      email,
+      source: id,
+      description: `Subscriber for ${userId}`
+    })
+
+    sub = await stripe.subscriptions.create({
+      customer: customer.id,
+      items: [
+        {
+          plan
+        }
+      ]
+    })
+  } catch (err) {
+    logger.error(err)
+    res.status(500).send('Unexpected error while submitting payment.')
+    return
+  }
+
+  let user
+  try {
+    user = await User.findOne({ id: userId })
+  } catch (err) {
+    logger.error(err)
+    res.status(401).send('Unexpected error while finding user.')
+    return
+  }
+
+  if (!user) {
+    res.status(404).send('Could not find user data.')
+  }
+
+  try {
+    const { data } = user
+    const now = new Date()
+    const newData = { ...data, subscribed: now, subscriptionId: sub.id, customerId: customer.id }
```

Comment:
What does the subscription ID tell us? Is that just an identifier to the plan in Stripe? If we have more than one tier/level/class of subscription, can we identify what tier the user is in from this ID?

We also have an array of user roles on the user data (https://github.com/streetmix/streetmix/blob/d3a4540c949ff13c503728724e6f2220da008419/app/models/user.js, https://github.com/streetmix/streetmix/blob/d3a4540c949ff13c503728724e6f2220da008419/app/data/user_roles.json) that we set up a few months ago. The thought was that we could give people arbitrary role categories (regardless of whether or not they used Stripe to subscribe!) and that role represents a set of feature flags that would be set or unset for them. We only have two roles active right now: a default user and an admin, but the plan was to be able to assign a role to each tier of subscriber.

Reply:
Ah yeah good point, we'll definitely want to include plan ID here as well. I'll add a tier1 subscriber role to complement.